File: backend/app/risk/scoring.py
"""
Vendor Risk Scoring Engine

Computes vendor risk using:
- Filing delay frequency
- Invoice mismatch ratio
- IRN absence ratio
- Tax payment default rate
- Graph centrality score (network risk)

Formula:
Risk Score = 0.30*FilingDelay + 0.25*MismatchRatio + 0.20*IRNMissing + 0.15*TaxDefault + 0.10*NetworkRisk

Also includes XGBoost predictive model for historical pattern matching.
"""

import numpy as np
import networkx as nx
from app.database import get_session


def compute_vendor_risk(gstin: str) -> dict:
    """Compute comprehensive risk score for a vendor using graph traversal."""
    with get_session() as session:
        # Component 1: Filing delay frequency
        filing_result = session.run("""
            MATCH (t:Taxpayer {gstin: $gstin})-[:FILES]->(r:Return)
            RETURN count(r) AS totalReturns,
                   count(CASE WHEN r.status = 'LATE' THEN 1 END) AS lateReturns,
                   count(CASE WHEN r.status = 'NOT_FILED' THEN 1 END) AS unfiledReturns
        """, gstin=gstin).single()
        
        total_returns = filing_result["totalReturns"] or 1
        late = filing_result["lateReturns"] or 0
        unfiled = filing_result["unfiledReturns"] or 0
        filing_delay = (late + unfiled * 2) / total_returns
        filing_delay = min(filing_delay, 1.0)
        
        # Component 2: Invoice mismatch ratio
        mismatch_result = session.run("""
            MATCH (t:Taxpayer {gstin: $gstin})-[:FILES]->(:Return {type: 'GSTR1'})-[:DECLARES]->(inv:Invoice)
            WITH count(inv) AS totalDeclared
            
            MATCH (t:Taxpayer {gstin: $gstin})-[:FILES]->(:Return {type: 'GSTR1'})-[:DECLARES]->(inv:Invoice)
            WHERE NOT EXISTS { MATCH (inv)-[:REFLECTED_IN]->(:Return {type: 'GSTR2B'}) }
            WITH totalDeclared, count(inv) AS notReflected
            
            RETURN totalDeclared, notReflected
        """, gstin=gstin).single()
        
        total_declared = mismatch_result["totalDeclared"] or 1
        not_reflected = mismatch_result["notReflected"] or 0
        mismatch_ratio = not_reflected / total_declared
        
        # Component 3: IRN absence ratio
        irn_result = session.run("""
            MATCH (t:Taxpayer {gstin: $gstin})-[:FILES]->(:Return {type: 'GSTR1'})-[:DECLARES]->(inv:Invoice)
            WITH count(inv) AS totalInvoices
            
            MATCH (t:Taxpayer {gstin: $gstin})-[:FILES]->(:Return {type: 'GSTR1'})-[:DECLARES]->(inv:Invoice)
            WHERE NOT EXISTS { MATCH (inv)-[:HAS_IRN]->(:IRN {status: 'ACTIVE'}) }
            WITH totalInvoices, count(inv) AS missingIrn
            
            RETURN totalInvoices, missingIrn
        """, gstin=gstin).single()
        
        total_invoices = irn_result["totalInvoices"] or 1
        missing_irn = irn_result["missingIrn"] or 0
        irn_missing_ratio = missing_irn / total_invoices
        
        # Component 4: Tax payment default (based on ledger offsets)
        tax_result = session.run("""
            MATCH (t:Taxpayer {gstin: $gstin})-[:CLAIMED_ITC]->(gstr3b:Return {type: 'GSTR3B'})
            OPTIONAL MATCH (gstr3b)-[:OFFSET_BY]->(l:LedgerEntry)
            RETURN count(gstr3b) AS totalFiled,
                   count(l) AS withLedger
        """, gstin=gstin).single()
        
        total_filed = tax_result["totalFiled"] or 1
        with_ledger = tax_result["withLedger"] or 0
        tax_default = 1.0 - (with_ledger / total_filed) if total_filed > 0 else 0.0
        
        # Component 5: Network risk (degree centrality approximation)
        network_risk = _compute_network_risk(gstin, session)
        
        # Weighted formula
        risk_score = (
            0.30 * filing_delay
            + 0.25 * mismatch_ratio
            + 0.20 * irn_missing_ratio
            + 0.15 * tax_default
            + 0.10 * network_risk
        )
        risk_score = round(min(risk_score, 1.0), 4)
        
        # Persist risk score
        session.run(
            "MATCH (t:Taxpayer {gstin: $gstin}) SET t.riskScore = $score",
            gstin=gstin,
            score=risk_score,
        )
        
        # Classify
        if risk_score >= 0.7:
            level = "HIGH"
            recommendation = "Flag for audit. Suspend ITC claims pending verification."
        elif risk_score >= 0.3:
            level = "MEDIUM"
            recommendation = "Monitor closely. Verify top mismatched invoices."
        else:
            level = "LOW"
            recommendation = "Standard compliance. No immediate action required."
        
        return {
            "gstin": gstin,
            "riskScore": risk_score,
            "riskLevel": level,
            "components": {
                "filingDelay": round(filing_delay, 4),
                "mismatchRatio": round(mismatch_ratio, 4),
                "irnMissingRatio": round(irn_missing_ratio, 4),
                "taxDefault": round(tax_default, 4),
                "networkRisk": round(network_risk, 4),
            },
            "recommendation": recommendation,
        }


def _compute_network_risk(gstin: str, session) -> float:
    """
    Compute network risk using graph centrality.
    High-risk suppliers connected to many risky buyers = higher network risk.
    """
    result = session.run("""
        MATCH (t:Taxpayer {gstin: $gstin})-[:FILES]->(:Return)-[:DECLARES]->(inv:Invoice)
        MATCH (buyer:Taxpayer {gstin: inv.buyerGstin})
        WHERE buyer.riskScore IS NOT NULL
        RETURN avg(buyer.riskScore) AS avgBuyerRisk,
               count(DISTINCT buyer) AS connectedBuyers
    """, gstin=gstin).single()
    
    avg_risk = result["avgBuyerRisk"] or 0.0
    connected = result["connectedBuyers"] or 0
    
    # Normalize: more connections to risky entities = higher network risk
    if connected == 0:
        return 0.0
    
    return min(avg_risk * min(connected / 10.0, 1.0), 1.0)


def compute_all_vendor_risks() -> list[dict]:
    """Compute risk scores for all taxpayers in the graph."""
    results = []
    with get_session() as session:
        taxpayers = session.run("MATCH (t:Taxpayer) RETURN t.gstin AS gstin")
        gstins = [r["gstin"] for r in taxpayers]
    
    for gstin in gstins:
        try:
            result = compute_vendor_risk(gstin)
            results.append(result)
        except Exception as e:
            results.append({"gstin": gstin, "error": str(e)})
    
    return results

File: backend/src/risk/scoring.ts
import db from '../database';

export interface VendorRisk {
    gstin: string;
    name: string;
    state: string;
    riskScore: number;
    riskLevel: 'LOW' | 'MEDIUM' | 'HIGH';
    components: {
        filingDelay: number;
        mismatchRatio: number;
        irnMissingRatio: number;
        taxDefault: number;
        networkRisk: number;
    };
    recommendation: string;
}

export async function calculateVendorRisk(gstin: string): Promise<VendorRisk> {
    const session = db.getSession();
    try {
        // 1. Filing Delay Ratio
        const filingData = await session.run(`
      MATCH (t:Taxpayer {gstin: $gstin})-[:FILES]->(r:Return {returnType: 'GSTR1'})
      WITH count(r) AS total, sum(CASE WHEN r.status = 'LATE' OR r.status = 'NOT_FILED' THEN 1 ELSE 0 END) AS delayed
      RETURN CASE WHEN total = 0 THEN 1.0 ELSE toFloat(delayed) / total END AS ratio
    `, { gstin });
        const filingDelay = filingData.records.length > 0 ? (filingData.records[0].get('ratio') || 0) : 1.0;

        // 2. Mismatch Ratio & IRN Missing
        const invData = await session.run(`
      MATCH (r:Return {returnType: 'GSTR1', gstin: $gstin})-[:DECLARES]->(i:Invoice)
      OPTIONAL MATCH (i)-[:HAS_IRN]->(irn:IRN)
      OPTIONAL MATCH (i)-[:REFLECTED_IN]->(r2b:Return {returnType: 'GSTR2B'})
      WITH count(i) AS total,
           sum(CASE WHEN irn IS NULL OR irn.status = 'CANCELLED' THEN 1 ELSE 0 END) AS irn_missing,
           sum(CASE WHEN r2b IS NULL THEN 1 ELSE 0 END) AS mismatch
      RETURN 
        CASE WHEN total = 0 THEN 0.0 ELSE toFloat(mismatch) / total END AS mismatch_ratio,
        CASE WHEN total = 0 THEN 0.0 ELSE toFloat(irn_missing) / total END AS irn_missing_ratio
    `, { gstin });

        const mismatchRatio = invData.records.length > 0 ? (invData.records[0].get('mismatch_ratio') || 0) : 0;
        const irnMissingRatio = invData.records.length > 0 ? (invData.records[0].get('irn_missing_ratio') || 0) : 0;

        // 3. Tax Default Indicator (Mock component for scope)
        const taxDefault = filingDelay > 0.8 ? 0.7 : 0.1;

        // 4. Network Risk (Derived purely via Cypher instead of NetworkX)
        // Identifies if supplier is connected to other high-risk buyers
        const netRiskData = await session.run(`
      MATCH (s:Taxpayer {gstin: $gstin})-[:FILES]->(:Return)-[:DECLARES]->(i:Invoice)
      MATCH (i)-[:REFLECTED_IN]->(:Return)<-[:FILES]-(b:Taxpayer)
      WHERE b.riskScore > 0.7
      RETURN count(DISTINCT b) AS risky_connections
    `, { gstin });
        const riskyConnections = netRiskData.records.length > 0 ? (netRiskData.records[0].get('risky_connections') || 0) : 0;
        const networkRisk = Math.min(Number(riskyConnections) * 0.2, 1.0);

        // Calculate Final Weighted Score
        const riskScore = (
            0.30 * filingDelay +
            0.25 * mismatchRatio +
            0.20 * irnMissingRatio +
            0.15 * taxDefault +
            0.10 * networkRisk
        );

        let riskLevel: 'LOW' | 'MEDIUM' | 'HIGH' = 'LOW';
        let recommendation = 'Standard compliance.';

        if (riskScore >= 0.7) {
            riskLevel = 'HIGH';
            recommendation = 'Flag for audit. Suspend ITC claims pending verification.';
        } else if (riskScore >= 0.3) {
            riskLevel = 'MEDIUM';
            recommendation = 'Monitor closely. Verify top mismatched invoices.';
        }

        // Persist to DB
        await session.run(`
      MATCH (t:Taxpayer {gstin: $gstin})
      SET t.riskScore = $score
      RETURN t.name AS name, t.state AS state
    `, { gstin, score: riskScore });

        const nameQuery = await session.run(`MATCH (t:Taxpayer {gstin: $gstin}) RETURN t.name AS name, t.state AS state`, { gstin });
        const name = nameQuery.records[0]?.get('name') || "Unknown";
        const state = nameQuery.records[0]?.get('state') || "Unknown";

        return {
            gstin,
            name,
            state,
            riskScore,
            riskLevel,
            components: {
                filingDelay,
                mismatchRatio,
                irnMissingRatio,
                taxDefault,
                networkRisk
            },
            recommendation
        };

    } finally {
        await session.close();
    }
}
